compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/gc/shared/GCUtils.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.hotspot.gc.shared;
+
+import jdk.vm.ci.meta.JavaKind;
+import org.graalvm.compiler.nodes.memory.AbstractWriteNode;
+import org.graalvm.compiler.nodes.memory.FixedAccessNode;
+import org.graalvm.compiler.nodes.memory.HeapAccess;
+
+/**
+ * Collection of GC related utility methods.
+ */
+public final class GCUtils {
+
+    /**
+     * Enum indicating the required precision of barriers.
+     */
+    public enum BarrierPrecision {
+        NONE,
+        PRECISE,
+        IMPRECISE
+    }
+
+    /**
+     * Given a node and its BarrierType, determine the required precision of barriers.
+     *
+     * @param node the node for which to determine barrier precision
+     * @param useDeferredInitBarriers whether or not to use deferred initialization barriers
+     * @return the required barrier precision
+     */
+    public static BarrierPrecision getBarrierPrecision(FixedAccessNode node, boolean useDeferredInitBarriers) {
```

Comment:
This could be a static method on enum `BarrierPrecision`, but it could just as easily be with the `BarrierType` code.

Reply:
Putting it in the enum sounds good. I'd rather not put it in the BarrierType class, because this is exposed everywhere.
I also have an idea how to reliably determine object vs. primitive. Stay tuned.